AK/StringView.h

```diff
@@ -38,6 +38,8 @@ class StringView {
 
     StringView substring_view(int start, int length) const;
     Vector<StringView> split_view(char) const;
+    StringView substring_view_from_substring(const StringView& substring) const;
```

Comment:
These two API's are extremely confusing. I think we can improve the names, how about `substring_view_starting_from_substring` and `substring_view_starting_after_substring`?

Another idea: how about adding a `StringView::offset()` and giving a default value to the `length` argument in `StringView::substring_view()` (where the default (0) means that it uses the remainder of the string) so you could do something like:

```cpp
view.substring_view(other_view.offset()); // same as view.substring_view_from_substring(other_view);
view.substring_view(other_view.offset() + other_view.length()); // same as view.substring_view_after_substring(other_view);
```

What do you think?

Reply:
`_*starting_*` does make it clearer, but that's a mouthful!

The default of 0 would clash with when you actually want to create a zero-length substring (perhaps not explicitly, but as a result of subtracting some length from something); lets use `-1`, or — I'm not sure if C++ allows it —
```cpp
StringView substring_view(int start, int length = this->length() - start) const;```

Do you mean for offset to be a third field, passed into (some of) the constructors by `substring_view()` implementation? Would that work with nested substrings? (I guess it could if we subtract `this->m_offset`...)

Reply:
You're right, `-1` makes more sense. I don't think C++ will let you put arbitrary expressions like that in a function declaration, sadly. I've wanted to do similar things many times ;)

Hm, damn,  I was thinking about a completely different substring class and forgot that StringView can't easily compute the "offset". So that idea won't work unless we require always passing a "base", which I don't think we should.

So hm.. I think maybe your original API is fine, but with the clearest names possible. I don't care if they're a mouthful, as long as we can understand them after we forget about them and then find them 6 months from today :)

Reply:
Fine, I'll change the names then, and I'd also try to write some comments & examples.

Speaking of which, is there any way to add unit tests (to AK)? And what does AK even stand for ("advanced kit")? And getting really off topic, why is there Apple copyright in `Function.h`?

Reply:
There's no straightforward way to add unit tests yet (but feel free to come up with one!)
The best way is probably to have unit tests that can run on the host system in a new /Tests/ directory. It would be nice to have the ability to run `valgrind` and such on them.

So the point of AK is to work both in kernel and userspace. The name was supposed to mean something like "agnostic kit" or "abstract kit" and AK are also my initials. I feel like there was some other meaning at some point but I forget. Nowadays I just pronounce it "ack" and stopped thinking about what it means. :)

The Apple copyright in Function.h is because that code was written by an old friend of mine when I worked at Apple. When I was bringing up the Serenity kernel I wanted to have something like std::function, but I couldn't compile the STL yet, so I had to make my own function class. I was in a hurry to make progress on something else, so I didn't want to get slowed down by figuring out how to implement a function object, so I borrowed the BSD-licensed Apple one (it's from the WebKit project btw :).)